server/src/main/java/io/druid/server/coordinator/helper/DruidCoordinatorRuleRunner.java

```diff
@@ -77,13 +84,46 @@ public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
       return params;
     }
 
+    // find available segments which are not overshadowed by other segments in DB
+    // only those would need to be loaded/dropped
+    // anything overshadowed by served segments is dropped automatically by DruidCoordinatorCleanupOvershadowed
+    Map<String, VersionedIntervalTimeline<String, DataSegment>> timelines = new HashMap<>();
+    for (DataSegment segment : params.getAvailableSegments()) {
+      VersionedIntervalTimeline<String, DataSegment> timeline = timelines.get(segment.getDataSource());
+      if (timeline == null) {
+        timeline = new VersionedIntervalTimeline<>(Comparators.comparable());
+        timelines.put(segment.getDataSource(), timeline);
+      }
+
+      timeline.add(
+          segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(segment)
+      );
+    }
+
+    Set<DataSegment> overshadowed = new HashSet<>();
+    for (VersionedIntervalTimeline<String, DataSegment> timeline : timelines.values()) {
+      for (TimelineObjectHolder<String, DataSegment> holder : timeline.findOvershadowed()) {
+        for (DataSegment dataSegment : holder.getObject().payloads()) {
+          overshadowed.add(dataSegment);
+        }
+      }
+    }
```

Comment:
Instead of filtering overshadowed in CoordinatorRuleRunner. 
We could also have a DruidCoordinatorFilterOvershadowed runnable that filters all overshadowed segments from available segments. I dont think we need overshadowed segments in any of the runnables. 
Also, with filtering of overshadowed segments, we can remove DruidCoordinatorCleanupOvershadowed as DruidCoordinatorCleanupUnneeded will cleanup overshadowed segments as well if they are filtered from available segments.